CLW/CL/CLW.cl

```diff
@@ -1492,3 +1507,196 @@ __kernel void segmented_distribute_part_sum_int_nocut(
+                                          __local type* shared_mem,\
+                                          __global type* out)\
+{\
+    int global_id = get_global_id(0);\
+    int group_id = get_group_id(0);\
+    int local_id = get_local_id(0);\
+    int group_size = get_local_size(0);\
+    \
+    if (global_id < buf_count)\
+    {\
+        assign_##type(shared_mem + local_id, buffer[global_id]);\
+    }\
+    else\
+    {\
+        assign_##type(shared_mem + local_id, neutral_##bin_op##_##type);\
+    }\
+    \
+    if (global_id == 0)\
+    {\
+        *out = neutral_##bin_op##_##type;\
+    }\
+    \
+    barrier(CLK_LOCAL_MEM_FENCE);\
+    for (int i = group_size / 2; i > 0; i >>= 1)\
+    {\
+        if (local_id < i)\
+        {\
+            assign_##type(shared_mem + local_id,\
+                          bin_op(shared_mem[local_id], shared_mem[local_id + i]));\
+        }\
+        barrier(CLK_LOCAL_MEM_FENCE);\
+    }\
+    \
+    if (local_id == 0)\
+    {\
+        atomic_##bin_op##_##type(out, shared_mem[0]);\
+    }\
+}
+
+// --------------------- NORMALIZATION ------------------------
+
+#define DEFINE_BUFFER_NORMALIZATION(type)\
+__kernel void buffer_normalization_##type(__global type* input,\
+                                          __global type* output,\
+                                          int buffer_count,\
+                                          __local type* shared_mem,\
+                                          __global type* auxiliary_buf)\
+{\
+    int global_id = get_global_id(0);\
+    int group_id = get_group_id(0);\
+    int local_id = get_local_id(0);\
+    int group_size = get_local_size(0);\
+    \
+    __local type* min_buffer = shared_mem;\
+    __local type* max_buffer = shared_mem + group_size;\
+    \
+    if (global_id < buffer_count)\
+    {\
+        min_buffer[local_id] = input[global_id];\
+        max_buffer[local_id] = input[global_id];\
+    }\
+    else\
+    {\
+        min_buffer[local_id] = neutral_min_##type;\
+        max_buffer[local_id] = neutral_max_##type;\
+    }\
+    \
+    if (global_id == 0)\
+    {\
+        auxiliary_buf[0] = neutral_min_##type;\
+        auxiliary_buf[1] = neutral_max_##type;\
+    }\
+    \
+    barrier(CLK_LOCAL_MEM_FENCE);\
+    \
+    for (int i = group_size / 2; i > 0; i >>= 1)\
+    {\
+        if (local_id < i)\
+        {\
+            assign_##type(min_buffer + local_id,\
+                          min(min_buffer[local_id], min_buffer[local_id + i]));\
+            assign_##type(max_buffer + local_id,\
+                          max(max_buffer[local_id], max_buffer[local_id + i]));\
+        }\
+        barrier(CLK_LOCAL_MEM_FENCE);\
+    }\
+    \
+    if (local_id == 0)\
+    {\
+        atomic_min_##type(auxiliary_buf, min_buffer[0]);\
+        atomic_max_##type(auxiliary_buf + 1, max_buffer[0]);\
+    }\
+    \
+    barrier(CLK_LOCAL_MEM_FENCE);\
+    type diff = auxiliary_buf[1] - auxiliary_buf[0];\
```

Comment:
again, global sync is required, not going to work

Reply:
Rewrited